pkg/conditions/replicasready/update.go

```diff
@@ -0,0 +1,47 @@
+package replicasready
+
+import (
+	capi "sigs.k8s.io/cluster-api/api/v1alpha3"
+	capiexp "sigs.k8s.io/cluster-api/exp/api/v1alpha3"
+	capiconditions "sigs.k8s.io/cluster-api/util/conditions"
+)
+
+func update(machinePool *capiexp.MachinePool) {
+	// If value is not specified when MachinePool CR is created, the default
+	// value is ensured in azure-admission-controller.
+	desiredReplicas := *machinePool.Spec.Replicas
```

Comment:
Is it really ok to trust the admission controller?
We might remove the defaulting any time and this line of code would panic.
I'd add a nil check.

Reply:
`MachinePool.Spec.Replicas`, when not specified, have value of `1` by the Cluster API spec, so we really should not remove that defaulting.

See here https://github.com/kubernetes-sigs/cluster-api/blob/master/exp/api/v1alpha3/machinepool_types.go#L39-L41

Upstream Cluster API project is also defaulting the value to `1` (per its own spec):
https://github.com/kubernetes-sigs/cluster-api/blob/v0.3.12/exp/api/v1alpha3/machinepool_webhook.go#L50-L51

and then it is used without any checks, e.g. see:
https://github.com/kubernetes-sigs/cluster-api/blob/dadd7bd6233edb1e422e7af282e923abff52f830/exp/controllers/machinepool_controller_phases.go#L67

which IMHO makes sense, as there would be little point to do the validation if then every usage of the field in the code has to do the same check.

A panic is probably more suitable in this case, it would surface more quickly, as returning error could hide (for at least some time) a serious problem.